server/src/main/java/org/elasticsearch/index/mapper/LuceneDocument.java

```diff
@@ -104,6 +105,33 @@ public void onlyAddKey(Object key, IndexableField field) {
         keyedFields.put(key, field);
     }
 
+    /**
+     * Add fields so that they can later be fetched using {@link #getByKey(Object)}.
+     * Allows one to provide a function for dealing with conflicting features if the key already exists.
+     */
+    public void addOrUpdateWithKey(
+        Object key,
+        IndexableField field,
+        BiFunction<IndexableField, IndexableField, IndexableField> remappingFunction
+    ) {
+        if (keyedFields == null) {
+            keyedFields = new HashMap<>();
+        }
+
+        if (keyedFields.containsKey(key)) {
+            IndexableField currentField = keyedFields.get(key);
+            IndexableField updatedField = remappingFunction.apply(currentField, field);
+            if (currentField != updatedField) {
+                keyedFields.put(key, updatedField);
+                fields.remove(currentField);
```

Comment:
What is the performance impact here? Its common to have 100s of fields, and those could have 100s of values. To my knowledge, `remove` does a linear scan of the values calling `equals` on each one. This ultimately just does an object identity check (as nobody overrides the `equals` method). 

One thought I had was to have the `keyedFields` actually be a map of
```
Map<Object, IndexableFieldAndPos> keyedFields
...
private record IndexableFieldAndPos(IndexableField value, int pos){}
```

This way we keep track of the position in the indexable list. This keeps us from doing two things:

 - Iterating the entire list on remove
 - Accidentally re-ordering the list (if somebody else uses this method in the future).

Reply:
I'm glad you brought it up; I had some local notes to discuss with you on this.  Seemed like there were a couple of structures that could be updated either keyedFIelds as you suggest or the fields list could changed to be a Map like this this: 
`Map<Integer, IndexableField> fields` 
where the Integer here is a hashcode.  

Your suggestion is straightforward though I'll roll with that for now.

There's some weirdness in this class in general related to getting fields with the same name that I don't fully understand which might be worth further review.